misc/rb.c

```diff
@@ -0,0 +1,495 @@
+			 * from there */
+			stacksz -= 2;
+			continue;
+		}
+
+		/* We can rotate locally to fix the whole tree.  First
+		 * make sure that node is on the same side of parent
+		 * as parent is of grandparent.
+		 */
+		int parent_side = get_side(parent, node);
+
+		if (parent_side != side) {
+			rotate(stack, stacksz);
+			node = stack[stacksz - 1];
+		}
+
+		/* Rotate the grandparent with parent, swapping colors */
+		rotate(stack, stacksz - 1);
+		set_color(stack[stacksz - 3], BLACK);
+		set_color(stack[stacksz - 2], RED);
+		return;
+	}
+
+	/* If we exit the loop, it's because our node is now the root,
+	 * which must be black */
+	set_color(stack[0], BLACK);
+}
+
+void rb_insert(struct rbtree *tree, struct rbnode *node)
+{
+	set_child(node, 0, NULL);
+	set_child(node, 1, NULL);
+
+	if (!tree->root) {
+		tree->root = node;
+		tree->max_depth = 1;
+		set_color(node, BLACK);
+		return;
+	}
+
+	struct rbnode *stack[tree->max_depth + 1];
+
+	int stacksz = find_and_stack(tree, node, stack);
+
+	struct rbnode *parent = stack[stacksz - 1];
+
+	int side = tree->lessthan_fn(parent, node);
+
+	set_child(parent, side, node);
+	set_color(node, RED);
+
+	stack[stacksz++] = node;
+	fix_extra_red(stack, stacksz);
+
+	if (stacksz > tree->max_depth) {
+		tree->max_depth = stacksz;
+	}
+
+	/* We may have rotated up into the root! */
+	tree->root = stack[0];
+	CHECK(is_black(tree->root));
+}
+
+/* Called for a node N (at the top of the stack) which after a
+ * deletion operation is "missing a black" in its subtree.  By
+ * construction N must be black (because if it was red it would be
+ * trivially fixed by recoloring and we wouldn't be here).  Fixes up
+ * the tree to preserve red/black rules.  The "null_node" pointer is
+ * for situations where we are removing a childless black node.  The
+ * tree munging needs a real node for simplicity, so we use it and
+ * then clean it up (replace it with a simple NULL child in the
+ * parent) when finished.
+ */
+static void fix_missing_black(struct rbnode **stack, int stacksz,
+			      struct rbnode *null_node)
+{
+	/* Loop upward until we reach the root */
+	while (stacksz > 1) {
+		struct rbnode *c0, *c1, *inner, *outer;
+		struct rbnode *n = stack[stacksz - 1];
+		struct rbnode *parent = stack[stacksz - 2];
+		int n_side = get_side(parent, n);
+		struct rbnode *sib = get_child(parent, !n_side);
+
+		CHECK(is_black(n));
+
+		/* Guarantee the sibling is black, rotating N down a
+		 * level if needed (after rotate() our parent is the
+		 * child of our previous-sibling, so N is lower in the
+		 * tree)
+		 */
+		if (!is_black(sib)) {
+			stack[stacksz - 1] = sib;
+			rotate(stack, stacksz);
+			set_color(parent, RED);
+			set_color(sib, BLACK);
+			stack[stacksz++] = n;
+
+			parent = stack[stacksz - 2];
+			sib = get_child(parent, !n_side);
+		}
+
+		CHECK(sib);
+
+		/* Cases where the sibling has only black children
+		 * have simple resolutions
+		 */
+		c0 = get_child(sib, 0), c1 = get_child(sib, 1);
```

Comment:
curious why the comma operator used instead of just putting each on a line?

Reply:
Because it was a cut and paste of a previous declaration and I didn't change the syntax completely.   Will fix.